pkg/controller/control.go

```diff
@@ -1244,29 +1248,153 @@ func (c *Controller) handleErrorNoLock(err error) error {
+//
+//	return a list of replicas that need to be set to ERR on ENOSPC error.
+//	For example:
+//	replicaToErrOnNoSpaceMap are {"replica A": 1, "replica B": 2, "replica C": 1, "replica D": 2, "replica E": 0,}
+//	Written Bytes 2: {"replica B", "replica D"},
+//	Written Bytes 1: {"replica A", "replica C"},
+//	Written Bytes 0: {"replica E"}
+//	{"replica B", "replica D"} will be retained RW because they have max number of the replicas that wrote the same written bytes, and the written bytes is maximum.
+//	Therefore, it will return the list {"replica A", "replica C", "replica E"} to be set to ERR.
+func listReplicasToErrOnEnospc(replicaToErrOnNoSpaceMap map[string]int) []string {
+	writeBytesReplicasMap := getReplicaWriteBytesMap(replicaToErrOnNoSpaceMap)
+	maxWrittenBytes := getMaxWrittenBytesForMaxNumberReplicas(writeBytesReplicasMap)
+
+	// Retain replicas in writeBytesReplicasMap[maxWrittenBytes] RW
+	return getFilteredReplicasToErrList(replicaToErrOnNoSpaceMap, writeBytesReplicasMap[maxWrittenBytes])
+}
+
+// getReplicaWriteBytesMap converts a map of replicas to write bytes into a map of write bytes to a set of replicas.
+//
+//	For Example:
+//		Input: replicasWriteBytesMap: {"replica A": 1, "replica B": 2, "replica C": 2, "replica D": 2, "replica E": 0,}
+//		Output: {2: {"replica B": true, "replica C": true, "replica D": true}, 0: {"replica E": true}, 1: {"replica A": true}}
+func getReplicaWriteBytesMap(replicasWriteBytesMap map[string]int) map[int]map[string]bool {
+	writeBytesReplicasMap := make(map[int]map[string]bool)
+
+	for address, wb := range replicasWriteBytesMap {
+		if writeBytesReplicasMap[wb] == nil {
+			writeBytesReplicasMap[wb] = make(map[string]bool)
+		}
+		writeBytesReplicasMap[wb][address] = true
+	}
+	return writeBytesReplicasMap
+}
+
+// getMaxWrittenBytesForMaxNumberReplicas returns the maximum written bytes for the maximum number of replicas.
+//
+//	For Example:
+//		Input: writeBytesReplicasMap: {2: {"replica B": true, "replica C": true,}, 0: {"replica D": true, "replica E": true}, 1: {"replica A": true}}
+//		Output: 2
+//		Because 2 replicas (B, C) is the maximum number of replicas and 2 bytes written is the maximum written byte (bigger then 0).
+func getMaxWrittenBytesForMaxNumberReplicas(writeBytesReplicasMap map[int]map[string]bool) int {
+	maxReplicasNum := 0
+	maxReplicasNumWriteBytesList := []int{}
+	for wb, addresses := range writeBytesReplicasMap {
+		if len(addresses) > maxReplicasNum {
+			maxReplicasNum = len(addresses)
+			maxReplicasNumWriteBytesList = []int{wb}
+			continue
+		}
+		if len(addresses) == maxReplicasNum {
+			maxReplicasNumWriteBytesList = append(maxReplicasNumWriteBytesList, wb)
+		}
+	}
+
+	maxWrittenBytes := int(-1)
+	for _, wb := range maxReplicasNumWriteBytesList {
+		if wb > maxWrittenBytes {
+			maxWrittenBytes = wb
+		}
+	}
+
+	return maxWrittenBytes
+}
+
+func getFilteredReplicasToErrList(replicasToErrMap map[string]int, retainReplicasRWMap map[string]bool) []string {
```

Comment:
nit: `retainedReplicas` seems no need the bool value.

```suggestion
func getFilteredReplicasToErrList(replicasToErrMap map[string]int, retainedReplicas map[string]struct{}) []string {
```

Reply:
Modified. Thanks.